niftynet/layer/convolution.py

```diff
@@ -220,3 +244,102 @@ def activation(output_tensor):
             output_tensor = activation(conv_layer(input_tensor))
 
         return output_tensor
+
+
+def _compute_pad_size(input_dim_size, output_dim_size, kernel_dim_size,
+                      stride, dilation):
+    """
+    Computes the size of the pad using the formula given in TF's conv_ops.cc.
+    :return: the one-sided pad size
+    """
+
+    return ((output_dim_size - 1)*stride + (kernel_dim_size - 1)*dilation + 2
+            - input_dim_size)//2
+
+
+def _extended_convolution(input_tensor,
+                          kernel,
+                          strides,
+                          dilations,
+                          padding,
+                          constant=0,
+                          name='extended_convolution'):
+    """
+    A simple wrapper for tf.nn.convolution that first expands the input tensor
+    by sampling at discrete locations in the original tensor then invokes
+    the original convolution operation on the expanded tensor, and finally
+    extracts a suitable output tensor from the output of the convolution of
+    the expanded tensor.
+    :param input_tensor: original convolution input tensor
+    :param kernel: convolution kernel
+    :param strides: strided convolution strides (one per spatial dimension)
+    :param dilations: dilated convolution dilation factors
+    (one per spatial dimension)
+    :param padding: a string specifying the type of padding to apply
+    :param constant: a padding constant (only read in the case of constant
+    padding)
+    :param name: a name for the operation
+    :return: a convolution result of the same size as the input tensor
+    """
+
+    input_shape = input_tensor.shape.as_list()
+    kernel_shape = kernel.shape.as_list()
+
+    if any(i is None or i < 0 or k is None or k < 0
+           for i, k in zip(input_shape, kernel_shape)):
+        raise ValueError('The dimensions of the input tensor and the filter'
+                         ' must be known in advance for this operation to '
+                         'work.')
+
+    output_shape = [int(math.ceil(i/s)) for i, s in
+                    zip(input_shape[1:-1], strides)]
+    output_shape = [input_shape[0]] + output_shape + [kernel_shape[-1]]
+
+    dimpads = [0]
+    for i, k, s, d in zip(input_shape[1:-1], kernel_shape[:-1],
+                          strides, dilations):
+        pad = _compute_pad_size(i, int(math.ceil(i/s)), k, s, d)
+        dimpads.append(pad)
+    dimpads += [0]
+
+    # Cannot pad by more than 1 dimension size => repeatedly pad
+    if padding in ('REFLECT', 'SYMMETRIC'):
+        padded_input = input_tensor
+        offset = 1 if padding == 'REFLECT' else 0
+
+        while min(o - i - 2*p for o, i, p in zip(
+                padded_input.shape.as_list(),
+                input_shape,
```

Comment:
the size of the batch dim (`input_shape[0]`) could be `None`, and it's never needed in this function except this line.  Could we replace `input_shape` with `input_shape[1:]`?

Reply:
Good point. I've adapted it the way you suggest in 32b3edaf70f9d94a753002557eb192e4c89603de.